storage2/src/state.rs

```diff
@@ -38,23 +41,74 @@ impl State {
 
 #[async_trait]
 impl StateRead for State {
-    fn get_raw(&self, key: &str) -> Result<Option<Vec<u8>>> {
+    async fn get_raw(&self, key: &str) -> Result<Option<Vec<u8>>> {
         // If the key is available in the unwritten_changes cache, return it.
         if let Some(v) = self.unwritten_changes.get(key) {
             return Ok(v.clone());
         }
 
         // Otherwise, if the key is available in the snapshot, return it.
-        self.snapshot.get_raw(key)
+        self.snapshot.get_raw(key).await
     }
 
-    fn get_sidecar(&self, key: &[u8]) -> Result<Option<Vec<u8>>> {
-        // If the key is available in the sidecar cache, return it.
-        if let Some(v) = self.sidecar_changes.get(key) {
+    async fn get_nonconsensus(&self, key: &[u8]) -> Result<Option<Vec<u8>>> {
+        // If the key is available in the nonconsensus cache, return it.
+        if let Some(v) = self.nonconsensus_changes.get(key) {
             return Ok(v.clone());
         }
 
         // Otherwise, if the key is available in the snapshot, return it.
-        self.snapshot.get_sidecar(key)
+        self.snapshot.get_nonconsensus(key).await
+    }
+
+    async fn prefix_raw(
+        &self,
+        prefix: &str,
+    ) -> Result<Pin<Box<dyn Stream<Item = (String, Box<[u8]>)> + Send + '_>>> {
+        // Interleave the unwritten_changes cache with the snapshot.
+        let (tx, rx) = mpsc::channel(100);
+
+        let mut snapshotted_stream = self.snapshot.prefix_raw(prefix).await?;
+        let mut snapshotted_match = snapshotted_stream.next().await;
+        for (key, value) in self.unwritten_changes.iter() {
+            // Iterate the unwritten_changes cache (sorted by key) until we reach the keys
+            // that match the prefix.
+            //
+            // If value is `None`, then the key has been deleted, and we should skip it.
+            if !key.starts_with(prefix) || value.is_none() {
+                continue;
+            }
```

Comment:
You can use the [`BTreeMap::range`](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.range) method here to skip right to the key you want to start with, which means you don't have to scan all the changes, just the relevant ones.